### PR TITLE
Added : Survey modal

### PR DESCRIPTION
**Pull request by shivamsherje**
## Discription :
-- Added survey button on home view .
-- Added survey feedback modal by clicking on survey button.
-- Add a function to get data of survey from strapy.
-- Add a function to post feedback of survey in dyanamoDB AWS.

### Survey button on homeview:
Screenshot 2023-10-16 at 11 33 27 AM

### Survey modal :
Screenshot 2023-10-16 at 11 33 14 AM

### CODE REVIEW

File: listeners/actions/index.js [hunk 1]
```diff
@@ -1,7 +1,9 @@
 const { sampleActionCallback } = require('./sample-action');
 const { openDailyCheckinModalCallback } = require('./open-daily-checkin-action');
+const { openSurveyModalCallback } = require('./open-survey-action');
 
 module.exports.register = (app) => {
   app.action('sample_action_id', sampleActionCallback);
   app.action('open-daily-checkin-modal', openDailyCheckinModalCallback);
+  app.action('open-Survey-modal', openSurveyModalCallback);
```
Comment (shivamsherje) [inline listeners/actions/index.js]:
```suggestion
  app.action('open-survey-modal', openSurveyModalCallback);
```

File: listeners/actions/open-survey-action.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const getSurveyModal = require('../../user-interface/models/survey-modal');
+
+const openSurveyModalCallback = async ({ ack, body, client }) => {
+  await ack();
+  const SurveyModal = await getSurveyModal(body.user.name);
```
Comment (shivamsherje) [inline listeners/actions/open-survey-action.js]:
```suggestion
  const surveyModal = await getSurveyModal(body.user.name);
```

File: listeners/actions/open-survey-action.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  try {
+    await client.views.open({
+      trigger_id: body.trigger_id,
+      view: SurveyModal,
```
Comment (shivamsherje) [inline listeners/actions/open-survey-action.js]:
```suggestion
      view: surveyModal,
```

File: listeners/views/survey.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+const { SurveyFeedback } = require('../../services/aws/index');
+
+// eslint-disable-next-line consistent-return
+const postSurveyFeedback = async ({ ack, body }) => {
+  await ack({
+    response_action: 'clear',
+  });
+  try {
+    const SurveyData = Object.values(body.view.state.values).map((value) => ({
```
Comment (shivamsherje) [inline listeners/views/survey.js]:
```suggestion
    const surveyData = Object.values(body.view.state.values).map((value) => ({
```

File: listeners/views/survey.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      question: Object.keys(value)[0],
+      answer: Object.values(value)[0].selected_option.value,
+    }));
+    SurveyFeedback({ data: SurveyData, body });
+    return SurveyData;
```
Comment (shivamsherje) [inline listeners/views/survey.js]:
```suggestion
    return surveyData;
```

File: listeners/views/survey.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+const { SurveyFeedback } = require('../../services/aws/index');
```
Comment (shivamsherje) [inline listeners/views/survey.js]:
```suggestion
const { surveyFeedback } = require('../../services/aws/index');
```

File: listeners/views/survey.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+const postSurveyFeedback = async ({ ack, body }) => {
+  await ack({
+    response_action: 'clear',
+  });
+  try {
+    const SurveyData = Object.values(body.view.state.values).map((value) => ({
+      question: Object.keys(value)[0],
+      answer: Object.values(value)[0].selected_option.value,
+    }));
+    SurveyFeedback({ data: SurveyData, body });
```
Comment (shivamsherje) [inline listeners/views/survey.js]:
```suggestion
    surveyFeedback({ data: surveyData, body });
```

File: services/strapi/index.js [hunk 8]
```diff
@@ -42,8 +42,26 @@ async function getDailyCheckinQuestions() {
     return [];
   }
 }
+async function getSurveyQuestions() {
+  try {
+    const url = `${baseUrl}/surveys?populate=*`;
+    const headers = {
+      Authorization: `Bearer ${process.env.SLACK_STRAPI_AUTH_TOKEN}`,
+    };
```
Comment (disha1202) [inline services/strapi/index.js]: Make this a global variable

File: services/aws/index.js [hunk 9]
```diff
@@ -36,4 +36,28 @@ async function dailyCheckinFeedback(params) {
   });
 }
 
-module.exports = { dailyCheckinFeedback };
+async function SurveyFeedback(params) {
+  const SurveysFeedback = params.data.map((item) => ({
```
Comment (shivamsherje) [inline services/aws/index.js]:
```suggestion
  const surveysFeedback = params.data.map((item) => ({
```

File: services/aws/index.js [hunk 10]
```diff
@@ -36,4 +36,28 @@ async function dailyCheckinFeedback(params) {
...
+    question: item.question,
+    answer: item.answer,
+  }));
+  const payload = {
+    TableName: 'Surveys',
+    Item: {
+      id: params.body.user.id + timestamp,
+      user_id: params.body.user.id,
+      timestamp,
+      SurveysFeedback,
```
Comment (shivamsherje) [inline services/aws/index.js]:
```suggestion
      surveysFeedback,
```

File: services/aws/index.js [hunk 11]
```diff
@@ -36,4 +36,28 @@ async function dailyCheckinFeedback(params) {
...
+  docClient.put(payload, (error) => {
+    if (error) {
+      console.error(error);
+    } else {
+      console.log('Daily checkins feedback posted successfully !!');
+    }
+  });
+}
+
+module.exports = { dailyCheckinFeedback, SurveyFeedback };
```
Comment (shivamsherje) [inline services/aws/index.js]:
```suggestion
module.exports = { dailyCheckinFeedback, surveyFeedback };
```

File: user-interface/models/survey-modal.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
+const { getSurveyQuestions } = require('../../services/strapi/index');
+
+async function getSurveyModal(username) {
+  const SurveyQuestions = await getSurveyQuestions();
```
Comment (disha1202) [inline user-interface/models/survey-modal.js]: Use camel case

Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
  const surveyQuestions = await getSurveyQuestions();
```

File: user-interface/models/survey-modal.js [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+  const questions = SurveyQuestions.map((SurveyQuestion) => ({
```
Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
  const questions = surveyQuestions.map((surveyQuestion) => ({
```

File: user-interface/models/survey-modal.js [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+      type: 'plain_text',
+      text: SurveyQuestion.question,
+      emoji: true,
+    },
+  }));
+
+  return {
+    title: {
+      type: 'plain_text',
+      text: 'Workplace check-in',
```
Comment (disha1202) [inline user-interface/models/survey-modal.js]: Make the title dynamic

File: user-interface/models/survey-modal.js [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        emoji: true,
+      },
+      options: [
+        {
+          text: {
+            type: 'plain_text',
+            text: 'good',
+            emoji: true,
+          },
+          value: SurveyQuestion.options[0],
```
Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
          value: surveyQuestion.options[0],
```

File: user-interface/models/survey-modal.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        },
+        {
+          text: {
+            type: 'plain_text',
+            text: 'better',
+            emoji: true,
+          },
+          value: SurveyQuestion.options[1],
```
Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
          value: surveyQuestion.options[1],
```

File: user-interface/models/survey-modal.js [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+        },
+        {
+          text: {
+            type: 'plain_text',
+            text: 'best',
+            emoji: true,
+          },
+          value: SurveyQuestion.options[2],
```
Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
          value: surveyQuestion.options[2],
```

File: user-interface/models/survey-modal.js [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        },
+        {
+          text: {
+            type: 'plain_text',
+            text: 'wonderfull',
+            emoji: true,
+          },
+          value: SurveyQuestion.options[3],
```
Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
          value: surveyQuestion.options[3],
```

File: user-interface/models/survey-modal.js [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+        },
+        {
+          text: {
+            type: 'plain_text',
+            text: 'Excieting',
+            emoji: true,
+          },
+          value: SurveyQuestion.options[4],
```
Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
          value: surveyQuestion.options[4],
```

File: user-interface/models/survey-modal.js [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+        },
+      ],
+      action_id: SurveyQuestion.question,
```
Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
      action_id: surveyQuestion.question,
```

File: user-interface/models/survey-modal.js [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+    },
+    label: {
+      type: 'plain_text',
+      text: SurveyQuestion.question,
```
Comment (shivamsherje) [inline user-interface/models/survey-modal.js]:
```suggestion
      text: surveyQuestion.question,
```

File: user-interface/models/survey-modal.js [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+  const questions = SurveyQuestions.map((SurveyQuestion) => ({
+    type: 'input',
+    element: {
+      type: 'static_select',
+      placeholder: {
+        type: 'plain_text',
+        text: 'I feel..',
+        emoji: true,
+      },
+      options: [
```
Comment (disha1202) [inline user-interface/models/survey-modal.js]: Make the options dynamic as well in surveys as well as daily checkin

**Review by disha1202 — CHANGES_REQUESTED** [review]

File: services/aws/index.js (hunk 9 above)
Comment (disha1202) [reply services/aws/index.js]:
```suggestion
  const surveyFeedback = params.data.map((item) => ({
```

File: listeners/actions/index.js (hunk 1 above)
Comment (shivamsherje) [reply listeners/actions/index.js]: done

File: listeners/actions/open-survey-action.js (hunk 2 above)
Comment (shivamsherje) [reply listeners/actions/open-survey-action.js]: done

File: listeners/actions/open-survey-action.js (hunk 3 above)
Comment (shivamsherje) [reply listeners/actions/open-survey-action.js]: done

File: listeners/views/survey.js (hunk 4 above)
Comment (shivamsherje) [reply listeners/views/survey.js]: done

File: listeners/views/survey.js (hunk 5 above)
Comment (shivamsherje) [reply listeners/views/survey.js]: done

File: listeners/views/survey.js (hunk 6 above)
Comment (shivamsherje) [reply listeners/views/survey.js]: done

File: listeners/views/survey.js (hunk 7 above)
Comment (shivamsherje) [reply listeners/views/survey.js]: done

File: services/aws/index.js (hunk 9 above)
Comment (shivamsherje) [reply services/aws/index.js]: done

File: services/aws/index.js (hunk 10 above)
Comment (shivamsherje) [reply services/aws/index.js]: done

File: services/aws/index.js (hunk 11 above)
Comment (shivamsherje) [reply services/aws/index.js]: done

File: user-interface/models/survey-modal.js (hunk 12 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 13 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 15 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 16 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 17 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 18 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 19 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 20 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 21 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: services/strapi/index.js (hunk 8 above)
Comment (shivamsherje) [reply services/strapi/index.js]: done

File: user-interface/models/survey-modal.js (hunk 12 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 14 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: user-interface/models/survey-modal.js (hunk 22 above)
Comment (shivamsherje) [reply user-interface/models/survey-modal.js]: done

File: services/aws/index.js [hunk 23]
```diff
@@ -36,4 +35,28 @@ async function dailyCheckinFeedback(params) {
...
+      timestamp: Math.floor(Date.now() / 1000),
+      feedback,
+    },
+  };
+
+  docClient.put(payload, (error) => {
+    if (error) {
+      console.error(error);
+    } else {
+      console.log('Daily checkins feedback posted successfully !!');
```
Comment (disha1202) [inline services/aws/index.js]: Improve the message

**Review by disha1202 — CHANGES_REQUESTED** [review]

**Review by disha1202 — APPROVED** [review]

**Pull request merged**